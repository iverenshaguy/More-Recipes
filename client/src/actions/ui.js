import { TOGGLE_MODAL, TOGGLE_REVIEW_FORM } from './actionTypes';

const toggleModal = modal => ({
  type: TOGGLE_MODAL,
  payload: modal || null
});


const toggleReviewForm = () => ({
  type: TOGGLE_REVIEW_FORM
});

export default {
  toggleModal,
  toggleReviewForm
};
