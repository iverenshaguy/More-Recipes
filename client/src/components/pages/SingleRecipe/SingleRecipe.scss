@import '../../../assets/scss/modules';

// .view-recipe #recipe-rating {
//   border-bottom: 1px solid #DFDFDF;
//   background-color: $white;
// }

#recipe-wrapper {
  border-radius: 0.5rem;
  border: 1px solid $white;
  background-color: $white;
  height: auto;

  h4 {
    font-weight: 300;
  }

  #recipe-rating {
    background-color: $white;
    border-bottom: 1px solid #DFDFDF;
  }

  #recipe-rating-info, #recipe-details-info, #recipe-preparation-info, #recipe-cooking-directions-info {
    float: right !important;
    text-align: right !important;
  }

  #ingredient-list p:before {
    font-family: Flaticon;
    content: "\f1d4";
    font-size: 1.2rem;
    color: $recipe-details-icon-grey;
  }
  
  #cooking-directions p:before {
    font-family: Flaticon;
    content: "\f1d9";
    display: inline-block;
    padding-right: 3px;
    vertical-align: middle;
    font-size: 1.2rem;
    color: $recipe-details-icon-grey;
  }
  
  #preparation p:before {
    font-family: Flaticon;
    content: "\f196";
    display: inline-block;
    padding-right: 3px;
    vertical-align: middle;
    font-size: 1.2rem;
    color: $recipe-details-icon-grey;
  }
  
  #ingredient-list p, #cooking-directions ol li {
    font-size: 1.1rem;
  }
  
  #recipe-actions {
    background-color: #EEEEEE;
    font-size: 1.3rem;
  
    p .fa {
      color: $recipe-details-icon-grey;
    }
  }
  
  .recipe-reaction-toolbar {
    width: 100%;
  
    button {
      width: 25%;
      padding: .75rem 1rem;
  
      &:hover, &:focus, &:active {
        background-color: $light-green !important;
        border-color: $light-green !important;
      }
    }
  }
  
  #side-rating-toolbar {
    position: fixed;
    top: 40%;
    left: 10%;
    display: none;
  
    .fa-star, .fa-star-o, .fa-heart, .fa-heart-o, .fa-thumbs-up, .fa-thumbs-o-up, .fa-thumbs-down, .fa-thumbs-o-down {
      color: $light-green;
    }
  }


  /*REVIEW WRAPPER*/
  
  #review-wrapper form p {
    font-size: 1.25rem;
  }
  
  #reviews img {
    height: 3rem;
    width: 3rem;
    border-radius: 50%;
  }
}

@media only screen and (max-width: 767px) {
  #recipe-wrapper {
    text-align: center;

    #recipe-rating-info, #recipe-details-info, #recipe-preparation-info, #recipe-cooking-directions-info {
      float: none !important;
    }
  }
}

@media only screen and (min-width: 880px) and (max-width: 992px) {
  #recipe-wrapper {
    #side-rating-toolbar {
      position: fixed;
      top: 40%;
      left: 18%;
    }
  }
}

/* CALL TO ACTION */
#call-to-action {
  color: $green !important;
  text-align: center;
  width: 100%;

  span {
    color: $green !important;
    width: 25% !important;

    &:hover {
      text-decoration: none;
    }
  }

  i {
    color: $green !important;

    &:hover {
      font-size: 1.6em;
    }
  }

  p {
    font-size: 1.1rem;
  }
}