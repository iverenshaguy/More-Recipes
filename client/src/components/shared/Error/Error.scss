@import '../../../assets/scss/modules';

#error-wrapper {
  background-color: $white;
  border: 1px solid $white;
  border-radius: 0.5rem;
  display: flex;
  height: 75vh;
}
