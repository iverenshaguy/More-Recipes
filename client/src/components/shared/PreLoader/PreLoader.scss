@import '../../../assets/scss/modules';

.loader {
  &-hidden {
    display: none;
  }
  
  &-active {
    display: block;
  }

  .ball-beat > div {
    background-color: $light-green;
  }

  .ball-scale-multiple {
    margin: 30vh 0 25vh 46vw;
    height: 11vh;
    width: 11vh;

    & > div {
      left: 0;
    }
  }
}