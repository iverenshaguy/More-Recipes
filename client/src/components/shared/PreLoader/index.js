import React from 'react';
import Loader from 'react-loaders';

const PreLoader = () => (<Loader type="semi-circle-spin" style={{ transform: 'scale(0.5)' }} />);

export default PreLoader;
