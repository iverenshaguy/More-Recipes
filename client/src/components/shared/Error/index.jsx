import React from 'react';
import PropTypes from 'prop-types';
import './Error.scss';

/**
 * @exports
 * @function Error
 * @param {object} props
 * @returns {JSX} Error
 */
const Error = (props) => {
  const { type } = props;
  return (
    <div className="container align-items-center justify-content-center" id="error-wrapper">
      <div className="container">
        <h3 className="text-center">{type === 404 ? 'Page Not Found' : 'Something happened, please try again'}</h3>
      </div>
    </div>
  );
};

Error.propTypes = {
  type: PropTypes.number
};

Error.defaultProps = {
  type: 500
};

export default Error;
