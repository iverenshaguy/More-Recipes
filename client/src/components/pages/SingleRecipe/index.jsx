import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Error from '../../shared/Error';
import PreLoader from '../../shared/PreLoader';
import { toggleReviewForm } from '../../../actions/ui';
import { voteRecipe } from '../../../actions/voteRecipe';
import SingleRecipeItem from '../../shared/SingleRecipeItem';
import { isReviewed } from '../../../selectors/recipeReviews';
import { fetchSingleRecipe } from '../../../actions/singleRecipe';
import { addRecipeToFavorites } from '../../../actions/favoriteRecipes';
import { recipeObjectPropTypes, urlMatchPropTypes, userPropTypes } from '../../../helpers/proptypes';
import './SingleRecipe.scss';

/**
 * @exports
 * @class SingleRecipe
 * @extends Component
 * @returns {JSX} SingleRecipe
 */
class SingleRecipe extends Component {
  static propTypes = {
    error: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.shape({
        recipeId: PropTypes.shape({
          msg: PropTypes.string
        })
      })
    ]),
    ...userPropTypes,
    ...urlMatchPropTypes,
    ...recipeObjectPropTypes,
    dispatch: PropTypes.func.isRequired,
    reviewed: PropTypes.bool.isRequired,
    isFetching: PropTypes.bool.isRequired,
  }

  static defaultProps = {
    error: null
  }

  /**
   * @constructor
   * @memberof SingleRecipe
   * @returns {nothing} Returns nothing
   */
  constructor() {
    super();
    this.toggleReviewForm = this.toggleReviewForm.bind(this);
    this.addRecipeToFavorites = this.addRecipeToFavorites.bind(this);
    this.voteRecipe = this.voteRecipe.bind(this);
  }

  /**
   * @memberof SingleRecipe
   * @return {nothing} Returns nothing
   */
  componentWillMount() {
    const { match } = this.props;
    const id = parseInt(match.params.id, 10);
    this.props.dispatch(fetchSingleRecipe(id));
  }

  /**
   * @memberof SingleRecipe
   * @returns {nothing} Returns nothing
   */
  toggleReviewForm() {
    this.props.dispatch(toggleReviewForm());
  }

  /**
   * @memberof SingleRecipe
   * @param {object} e
   * @returns {nothing} Returns nothing
   */
  addRecipeToFavorites(e) {
    e.preventDefault();
    const { recipe: { recipeItem: { id } } } = this.props;
    this.props.dispatch(addRecipeToFavorites(id));
  }

  /**
   * @memberof SingleRecipe
   * @param {object} e
   * @param {string} type
   * @returns {nothing} Returns nothing
   */
  voteRecipe(e, type) {
    e.preventDefault();
    const { recipe: { recipeItem: { id } } } = this.props;
    this.props.dispatch(voteRecipe(id, type));
  }

  /**
   * Renders SingleRecipe Component
   * @memberof SingleRecipe
   * @returns {JSX} SingleRecipe
   */
  render() {
    const {
      isFetching, error, recipe, user, reviewed
    } = this.props;

    return (
      <Fragment>
        {isFetching && <PreLoader />}
        {!isFetching && error && <Error type={500} />}
        {!isFetching && recipe && Object.keys(recipe).length !== 0 &&
          <div className="view-recipe" >
            <SingleRecipeItem
              user={user}
              recipe={recipe}
              reviewed={reviewed}
              voteRecipe={this.voteRecipe}
              toggleReviewForm={this.toggleReviewForm}
              addRecipeToFavorites={this.addRecipeToFavorites}
            />
          </div>}
      </Fragment>
    );
  }
}

const mapStateToProps = state => ({
  user: state.auth.user,
  isFetching: state.isFetching,
  recipe: state.singleRecipe.recipe.item,
  error: state.singleRecipe.recipe.error,
  reviewed: isReviewed(state.singleRecipe.recipeReviews.reviews, state.auth.user.id)
});

export { SingleRecipe as SingleRecipeComponent };

export default connect(mapStateToProps)(SingleRecipe);
