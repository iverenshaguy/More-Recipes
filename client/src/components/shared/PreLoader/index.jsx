import React from 'react';
import Loader from 'react-loaders';
import 'loaders.css/src/animations/ball-scale-multiple.scss';
import MiniPreLoader from './MiniPreLoader';
import './PreLoader.scss';

const PreLoader = () => (<Loader type="ball-scale-multiple" />);

export { MiniPreLoader };
export default PreLoader;
