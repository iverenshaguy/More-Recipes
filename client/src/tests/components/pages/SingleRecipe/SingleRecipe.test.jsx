import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import Error from '../../../../components/shared/Error';
import initialValues from '../../../setup/initialValues';
import { toggleReviewForm } from '../../../../actions/ui';
import PreLoader from '../../../../components/shared/PreLoader';
import SingleRecipeItem from '../../../../components/shared/SingleRecipeItem';
import { SingleRecipeComponent } from '../../../../components/pages/SingleRecipe';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const store = mockStore(initialValues);

const setup = () => {
  const props = {
    dispatch: jest.fn(),
    match: {
      url: '/',
      params: {
        id: 10
      }
    },
    isFetching: false,
    recipe: {
      recipeItem: {
        id: 2,
        recipeName: 'Recipe',
        userId: 1,
        isReviewed: false,
        ingredients: ['rice', 'beans'],
        preparations: ['wash', 'cook'],
        directions: ['wash', 'cook'],
        User: {
          id: 1,
          username: 'name'
        }
      }
    },
    error: null,
    user: {
      id: 2,
      username: 'username'
    },
    showReviewForm: false,
    reviewed: false
  };

  const shallowWrapper = shallow(<SingleRecipeComponent {...props} />);
  const mountedWrapper = mount(// eslint-disable-line
    <Provider store={store}>
      <SingleRecipeComponent {...props} />
    </Provider>);

  return { props, shallowWrapper, mountedWrapper };
};

describe('SingleRecipe', () => {
  it('renders correctly', () => {
    const { shallowWrapper } = setup();
    // Fix for component that returns React Fragment
    const fragment = shallowWrapper.instance().render();
    expect(shallow(<div>{fragment}</div>).getElement()).toMatchSnapshot();
  });

  it('renders Preloader when isFetching is true', () => {
    const { props } = setup();
    const shallowWrapper = shallow(<SingleRecipeComponent {...props} isFetching />);

    expect(shallowWrapper.find(PreLoader).length).toBeTruthy();
    expect(shallowWrapper.find(Error).length).toBeFalsy();
    expect(shallowWrapper.find(SingleRecipeItem).length).toBeFalsy();
  });

  it('renders Error when isFetching is false and there is an error', () => {
    const { props } = setup();
    const shallowWrapper = shallow(<SingleRecipeComponent
      {...props}
      isFetching={false}
      recipe={{}}
      error="An error occurred"
    />);

    expect(shallowWrapper.find(Error).length).toBeTruthy();
    expect(shallowWrapper.find(PreLoader).length).toBeFalsy();
    expect(shallowWrapper.find(SingleRecipeItem).length).toBeFalsy();
  });

  it('renders SingleRecipeItem when isFetching is false and there is a recipe', () => {
    const { props } = setup();
    const shallowWrapper = shallow(<SingleRecipeComponent
      {...props}
      isFetching={false}
    />);

    expect(shallowWrapper.find(SingleRecipeItem).length).toBeTruthy();
    expect(shallowWrapper.find(Error).length).toBeFalsy();
    expect(shallowWrapper.find(PreLoader).length).toBeFalsy();
  });

  it('calls toggleReviewForm() when review icon is clicked', () => {
    const { mountedWrapper, props } = setup();

    mountedWrapper.find('i.fa-star-o').simulate('click');

    expect(props.dispatch).toHaveBeenCalledWith(toggleReviewForm());
  });

  it('calls addRecipeToFavorites() when favorite icon is clicked', () => {
    const { mountedWrapper, props } = setup();

    mountedWrapper.find('i.fa-heart-o').simulate('click');

    expect(props.dispatch).toHaveBeenCalled();
  });

  it('calls voteRecipe() when upvote icon is clicked', () => {
    const { mountedWrapper, props } = setup();

    mountedWrapper.find('i.fa-thumbs-up').simulate('click');

    expect(props.dispatch).toHaveBeenCalled();
  });

  it('calls voteRecipe() when downvote icon is clicked', () => {
    const { mountedWrapper, props } = setup();

    mountedWrapper.find('i.fa-thumbs-down').at(1).simulate('click');

    expect(props.dispatch).toHaveBeenCalled();
  });
});
