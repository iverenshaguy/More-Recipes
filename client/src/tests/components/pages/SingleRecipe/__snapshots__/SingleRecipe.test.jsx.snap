// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleRecipe renders correctly 1`] = `
<div>
  <UNDEFINED>
    <div
      className="view-recipe"
    >
      <SingleRecipeItem
        addRecipeToFavorites={[Function]}
        error={null}
        recipe={
          Object {
            "recipeItem": Object {
              "User": Object {
                "id": 1,
                "username": "name",
              },
              "directions": Array [
                "wash",
                "cook",
              ],
              "id": 2,
              "ingredients": Array [
                "rice",
                "beans",
              ],
              "isReviewed": false,
              "preparations": Array [
                "wash",
                "cook",
              ],
              "recipeName": "Recipe",
              "userId": 1,
            },
          }
        }
        reviewed={false}
        toggleReviewForm={[Function]}
        user={
          Object {
            "id": 2,
            "username": "username",
          }
        }
        voteRecipe={[Function]}
      />
    </div>
  </UNDEFINED>
</div>
`;
